lossy_stream.go

```diff
@@ -0,0 +1,93 @@
+package webrtc
+
+import (
+	"fmt"
+	"io"
+	"sync"
+)
+
+// lossyReader wraps an io.Reader and discards data if it isn't read in time
+// Allowing us to only deliver the newest data to the caller
+type lossyReadCloser struct {
+	nextReader io.ReadCloser
+	mu         sync.RWMutex
+
+	incomingBuf chan []byte
+	amountRead  chan int
+
+	readError  error
+	hasErrored chan interface{}
+
+	closed chan interface{}
+}
+
+func newLossyReadCloser(nextReader io.ReadCloser) *lossyReadCloser {
+	l := &lossyReadCloser{
+		nextReader: nextReader,
+
+		closed: make(chan interface{}),
+
+		incomingBuf: make(chan []byte),
+		hasErrored:  make(chan interface{}),
+		amountRead:  make(chan int),
+	}
+
+	go func() {
+		readBuf := make([]byte, receiveMTU)
+		for {
+			i, err := nextReader.Read(readBuf)
+			if err != nil {
+				l.mu.Lock()
+				l.readError = err
+				l.mu.Unlock()
+
+				close(l.hasErrored)
+				break
+			}
+
+			select {
+			case in := <-l.incomingBuf:
+				copy(in, readBuf[:i])
+				l.amountRead <- i
+			default: // Discard if we have no inbound read
```

Comment:
Mind if we create an issue and do this in another PR, I don't think we can solve this short term without discussing a bit more (and it isn't a regression)

I think this ties into designing a good API + `rtpengine` not a hard problem, but I don't see us solving it in the next 48 hours

Reply:
I mean, if you're comfortable enough with the state of the code to merge, then sure go for it. I don't think there's a rush but it's your repo.

Reply:
I think i agree with Sean here, since it isn't a regression (it's new functionality), something that works (but may be sub-optimal) can always be improved upon later